packages/data-broker/src/services/dhis/DhisInstanceResolver.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { legacy_getDhisServerName } from '@tupaia/utils';
+
+export class DhisInstanceResolver {
+  constructor(models) {
+    this.models = models;
+  }
+
+  /**
+   * Find the DhisInstance to use. DhisInstanceResolver uses the [Standard Way], which is to have a
+   * DHIS instance marked against the DataSource. Otherwise, the DataSource might specify entityBasedDhisResolution,
+   * which then uses the provided entity code(s) to find the actual DhisInstance to use.
+   *
+   * @param {string|undefined} dataSourceDhisInstanceCode
+   * @param {string[]|undefined} entityCodes
+   * @return {Promise<DhisInstance>}
+   */
+  async get({ dataSourceDhisInstanceCode, entityCodes }) {
+    const entityBasedDhisResolution = entityCodes && entityCodes.length > 0;
+
+    if (!dataSourceDhisInstanceCode && !entityBasedDhisResolution) {
+      throw new Error(
+        'No DHIS Instance specified on dataSource, and entityBasedDhisResolution not used',
+      );
+    }
+
+    if (dataSourceDhisInstanceCode) {
+      const instance = await this.models.dhisInstance.findOne({
+        code: dataSourceDhisInstanceCode,
+      });
+      if (!instance) {
+        throw new Error(`No DHIS Instance found with code ${dataSourceDhisInstanceCode}`);
+      }
+      return instance;
+    }
+
+    return this.resolveMultipleEntityCodes(entityCodes);
+  }
+
+  /**
+   * @private
+   * @param {string} entityCode
+   * @return {Promise<DhisInstance>}
+   */
+  async resolveSingleEntityCode(entityCode) {
+    const serverName = await this.getDhisServerNameForEntity(entityCode);
+    return this.getFromServerName(serverName);
+  }
+
+  /**
+   * @private
+   * @param {string[]} entityCodes
+   * @return {Promise<DhisInstance>}
+   */
+  async resolveMultipleEntityCodes(entityCodes) {
+    const allDhisInstanceCodes = new Set();
+    const allDhisInstances = [];
+    for (const entityCode of entityCodes) {
+      const dhisInstance = await this.resolveSingleEntityCode(entityCode);
+      allDhisInstanceCodes.add(dhisInstance.code);
+      allDhisInstances.push(dhisInstance);
+    }
+    if (allDhisInstanceCodes.size > 1) {
+      throw new Error(
+        `All entities must use the same DHIS2 instance (got [${[...allDhisInstanceCodes].join(
+          ',',
+        )}])`,
+      );
+    }
+    return allDhisInstances[0];
```

Comment:
I left this as multiple single calls to legacy_getDhisServerName to make mocking easier. This way we can have a test in data-broker to make sure an error gets thrown if it resolves to multiple dhis servers. I thought about having this test in utils, on the legacy_getDhisServerName function, but I thought it might make more sense to have it on the non-legacy side of the codebase rather than the legacy side. I think the reason was that the legacy side is really hard to write tests for because mocking is hard. As this class should disappear once the data is cleaned up, are you happy to leave as is?

Reply:
Yup that sounds reasonable to me 👍